File: adapters/claude/src/agent_wake_claude/gating.py
"""HMAC gating for inbound wake events."""

import hmac
import hashlib


def verify_signature(body: bytes, secret: bytes, signature_header: str) -> bool:
    """Verify HMAC-SHA256 signature against the raw body.

    Args:
        body: The raw request body bytes.
        secret: The shared HMAC secret bytes.
        signature_header: The value of the X-AgentWake-Signature header,
            expected format: "sha256=<hex>".

    Returns:
        True if the signature is valid, False otherwise.
    """
    if not signature_header:
        return False

    expected = signature_header.strip()
    if not expected.startswith("sha256="):
        return False

    expected_hex = expected[7:]
    try:
        expected_digest = bytes.fromhex(expected_hex)
    except ValueError:
        return False

    computed = hmac.new(secret, body, hashlib.sha256).digest()
    return hmac.compare_digest(computed, expected_digest)

File: adapters/opencode.archived/src/gating.ts
import { createHmac } from "crypto";

export function verifySignature(
  body: Uint8Array,
  secret: Uint8Array,
  signatureHeader: string
): boolean {
  if (!signatureHeader) return false;

  const expected = signatureHeader.trim();
  if (!expected.startsWith("sha256=")) return false;

  const expectedHex = expected.slice(7);
  if (expectedHex.length !== 64) return false;

  const expectedBytes = new Uint8Array(32);
  for (let i = 0; i < 64; i += 2) {
    const byte = parseInt(expectedHex.substring(i, i + 2), 16);
    if (Number.isNaN(byte)) return false;
    expectedBytes[i / 2] = byte;
  }

  const hmac = createHmac("sha256", secret).update(body).digest();

  if (hmac.length !== expectedBytes.length) return false;
  let result = 0;
  for (let i = 0; i < hmac.length; i++) {
    result |= hmac[i] ^ expectedBytes[i];
  }
  return result === 0;
}
